src/dbnode/persist/schema/index_entry_hasher.go

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package schema
+
+import (
+	"hash"
+	"hash/adler32"
+	"sync"
+)
+
+type adlerHasher struct {
+	sync.Mutex
```

Comment:
So instead of using a hasher with a lock, we should just use `stackadler32` which was made specifically for zero alloc hashing:
https://github.com/m3db/stackadler32/blob/master/adler32.go

We already use this in the repo.

Best to just add a new function called `func HashIndexEntry` (which will be zero alloc and no locking):
```go
func (e IndexEntry) HashIDAndTags() int64 {
  hash := stackadler32.NewDigest().
    Update(e.ID).
    Update(e.EncodedTags)
  return hash.Sum32()
}
```

Also why is DataChecksum part of the checksum for just the index checksum? I thought we needed to differentiate between index checksum mismatches and the data checksum mismatches no?

Reply:
Hm as discussed, perhaps we can do a composite hash like we do in our map generators (if we want to combine datachecksum with the others):
```go
			hash := uint64(7)
			hash = 31*hash + xxhash.Sum64(e.ID)
			hash = 31*hash + xxhash.Sum64(e.EncodedTags)
			hash = 31*hash + uint64(e.DataChecksum)
```

Reply:
:+1: Sounds good, yeah :+1: will rename and update this hasher (may keep a single Hasher as discussed for testability)